File: py/yeriasdk/views/base_action_view.py
"""
Base action view class for ActionList and ActionGrid views
"""

from typing import List, Optional, Dict, Any, Callable
from abc import ABC, abstractmethod

from ..core.base_view import BaseView
from ..types.models import ActionConfig
from ..errors.exceptions import MissingRequiredParameterError


class BaseActionView(BaseView, ABC):
    """Abstract base class for action views (ActionList and ActionGrid)"""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._actions: List[ActionConfig] = []

    def add_action(
        self,
        code: str,
        title: str,
        description: Optional[str] = None,
        thumbnail: Optional[str] = None,
        disabled: bool = False,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> "BaseActionView":
        """Add an action"""
        if not code or not title:
            raise MissingRequiredParameterError("code and title")

        action = ActionConfig(
            code=code,
            title=title,
            desc=description,
            thumbnail=thumbnail,
            disabled=disabled,
            metadata=metadata,
        )

        self._actions.append(action)
        self._sync_actions_to_content()
        return self

    def add_actions(
        self,
        actions: List[Dict[str, Any]],
    ) -> "BaseActionView":
        """Add multiple actions at once"""
        for action in actions:
            self.add_action(
                code=action["code"],
                title=action["title"],
                description=action.get("description"),
                thumbnail=action.get("thumbnail"),
                disabled=action.get("disabled", False),
                metadata=action.get("metadata"),
            )
        return self

    def remove_action(self, action_code: str) -> bool:
        """Remove an action by code"""
        index = next(
            (i for i, action in enumerate(self._actions) if action.code == action_code),
            -1,
        )

        if index != -1:
            self._actions.pop(index)
            self._sync_actions_to_content()
            return True

        return False

    def update_action(
        self, action_code: str, updates: Dict[str, Any]
    ) -> bool:
        """Update an existing action"""
        action = next(
            (a for a in self._actions if a.code == action_code), None
        )

        if action:
            if "title" in updates:
                action.title = updates["title"]
            if "desc" in updates:
                action.desc = updates.get("desc")
            if "thumbnail" in updates:
                action.thumbnail = updates.get("thumbnail")
            if "disabled" in updates:
                action.disabled = updates.get("disabled", False)
            if "metadata" in updates:
                action.metadata = updates.get("metadata")
            self._sync_actions_to_content()
            return True

        return False

    def get_action(self, action_code: str) -> Optional[ActionConfig]:
        """Get an action by code"""
        return next((a for a in self._actions if a.code == action_code), None)

    def get_actions(self) -> List[ActionConfig]:
        """Get all actions"""
        return list(self._actions)

    def get_active_actions(self) -> List[ActionConfig]:
        """Get active (non-disabled) actions"""
        return [a for a in self._actions if not a.disabled]

    def get_action_count(self) -> int:
        """Get action count"""
        return len(self._actions)

    def has_action(self, action_code: str) -> bool:
        """Check if an action exists"""
        return any(a.code == action_code for a in self._actions)

    def clear_actions(self) -> None:
        """Clear all actions"""
        self._actions = []
        self._sync_actions_to_content()

    @abstractmethod
    def _sync_actions_to_content(self) -> None:
        """Abstract method to sync actions array to content"""
        pass

    def filter_actions(
        self, predicate: Callable[[ActionConfig], bool]
    ) -> List[ActionConfig]:
        """Filter actions by predicate"""
        return [a for a in self._actions if predicate(a)]

    def sort_actions(
        self, compare_fn: Callable[[ActionConfig, ActionConfig], int]
    ) -> "BaseActionView":
        """Sort actions by custom comparator"""
        self._actions.sort(key=lambda x: compare_fn(x, x))
        self._sync_actions_to_content()
        return self

    def sort_by_title(self, ascending: bool = True) -> "BaseActionView":
        """Sort actions by title"""
        self._actions.sort(key=lambda a: a.title, reverse=not ascending)
        self._sync_actions_to_content()
        return self

    def disable_all_actions(self) -> "BaseActionView":
        """Disable all actions"""
        for action in self._actions:
            action.disabled = True
        self._sync_actions_to_content()
        return self

    def enable_all_actions(self) -> "BaseActionView":
        """Enable all actions"""
        for action in self._actions:
            action.disabled = False
        self._sync_actions_to_content()
        return self

File: js/src/core/base/base-action-view.ts
import { BaseView } from '../base-view';
import { ActionConfig } from '../../types';
import { MissingRequiredParameterError } from '../../errors';

/**
 * Abstract subclass of BaseView shared by the "action" family of views
 * (ActionListView, ActionGridView, IconGridView).
 *
 * Factors out the action/item handling common to those views: adding,
 * updating, removing, querying, sorting and enabling/disabling the list of
 * actions that each concrete view renders. Subclasses implement
 * syncActionsToContent() to project the shared actions array into their own
 * content shape.
 *
 * This class is ABSTRACT and is never instantiated directly; use one of the
 * concrete action views instead.
 *
 * Base class for the action views (ActionList and ActionGrid)
 * Factors out the code common to ActionListView and ActionGridView
 */
export abstract class BaseActionView extends BaseView {
    protected actions: ActionConfig[] = [];

    /**
     * Adds an action
     */
    addAction(
        code: string,
        title: string,
        description?: string,
        thumbnail?: string,
        disabled: boolean = false,
        metadata?: Record<string, unknown>
    ): this {
        if (!code || !title) {
            throw new MissingRequiredParameterError('code and title');
        }

        const action: ActionConfig = {
            code,
            title,
            desc: description,
            thumbnail,
            disabled,
            metadata
        };

        this.actions.push(action);
        return this;
    }

    /**
     * Adds several actions at once
     */
    addActions(actions: Array<{
        code: string;
        title: string;
        description?: string;
        thumbnail?: string;
        disabled?: boolean;
        metadata?: Record<string, unknown>;
    }>): this {
        actions.forEach(action => {
            this.addAction(
                action.code,
                action.title,
                action.description,
                action.thumbnail,
                action.disabled,
                action.metadata
            );
        });
        return this;
    }

    /**
     * Removes an action by its code
     */
    removeAction(actionCode: string): boolean {
        const index = this.actions.findIndex(action => action.code === actionCode);

        if (index !== -1) {
            this.actions.splice(index, 1);
            return true;
        }

        return false;
    }

    /**
     * Updates an existing action
     */
    updateAction(actionCode: string, updates: Partial<ActionConfig>): boolean {
        const action = this.actions.find(a => a.code === actionCode);

        if (action) {
            Object.assign(action, updates);
            return true;
        }

        return false;
    }

    /**
     * Gets an action by its code
     */
    getAction(actionCode: string): ActionConfig | undefined {
        return this.actions.find(action => action.code === actionCode);
    }

    /**
     * Gets all actions
     */
    getActions(): ActionConfig[] {
        return [...this.actions];
    }

    /**
     * Gets the active actions (not disabled)
     */
    getActiveActions(): ActionConfig[] {
        return this.actions.filter(action => !action.disabled);
    }

    /**
     * Gets the number of actions
     */
    getActionCount(): number {
        return this.actions.length;
    }

    /**
     * Checks whether an action exists
     */
    hasAction(actionCode: string): boolean {
        return this.actions.some(action => action.code === actionCode);
    }

    /**
     * Clears all actions
     */
    clearActions(): void {
        this.actions = [];
    }

    /**
     * Abstract method to sync actions array to content
     * Child classes must implement this to update their specific content type
     */
    protected abstract syncActionsToContent(): void;

    /**
     * Filters the actions by a predicate
     */
    filterActions(predicate: (action: ActionConfig) => boolean): ActionConfig[] {
        return this.actions.filter(predicate);
    }

    /**
     * Sorts the actions by a custom comparator
     */
    sortActions(compareFn: (a: ActionConfig, b: ActionConfig) => number): this {
        this.actions.sort(compareFn);
        this.syncActionsToContent();
        return this;
    }

    /**
     * Sorts the actions by title (alphabetical order)
     */
    sortByTitle(ascending: boolean = true): this {
        return this.sortActions((a, b) => {
            const comparison = a.title.localeCompare(b.title);
            return ascending ? comparison : -comparison;
        });
    }

    /**
     * Disables all actions
     */
    disableAllActions(): this {
        this.actions.forEach(action => {
            action.disabled = true;
        });
        return this;
    }

    /**
     * Enables all actions
     */
    enableAllActions(): this {
        this.actions.forEach(action => {
            action.disabled = false;
        });
        return this;
    }
}
